orca-queue/src/main/kotlin/com/netflix/spinnaker/orca/q/handler/RunTaskHandler.kt

```diff
@@ -183,13 +186,24 @@ open class RunTaskHandler
     getContext()["continuePipeline"] == true
 
   private fun Stage<*>.withMergedContext(): Stage<*> {
-    // TODO: this isn't ideal as the additional data will get permanently added to the stage context
-    val context = getExecution().getContext().toMutableMap()
-    context.putAll(getContext())
-    this.setContext(context)
+    val execution = this.getExecution()
+    val transformed = contextParameterProcessor.process(getContext(), mapWithFallback(getContext(), execution), true)
+    setContext(mapWithFallback(transformed, execution))
```

Comment:
parses expressions in the stage context and provides necessary fallbacks to global context and refs to `trigger` or `execution`.